wandb/sdk/internal/writer.py

```diff
@@ -1,44 +1,200 @@
+        self._telemetry_overflow = False
 
-    def open(self):
+    def open(self) -> None:
         self._ds = datastore.DataStore()
         self._ds.open_for_write(self._settings.sync_file)
+        self._flow_control = flow_control.FlowControl(
+            settings=self._settings,
+            write_record=self._write_record,
+            forward_record=self._forward_record,
+            pause_marker=self._pause_marker,
+            recover_records=self._recover_records,
+        )
+
+    def _forward_record(self, record: "pb.Record") -> None:
+        self._context_keeper.add_from_record(record)
+        tracelog.log_message_queue(record, self._sender_q)
+        self._sender_q.put(record)
+
+    def _send_mark(self) -> None:
+        sender_mark = pb.SenderMarkRequest()
+        record = self._interface._make_request(sender_mark=sender_mark)
+        self._forward_record(record)
+
+    def _maybe_update_telemetry(self) -> None:
+        if self._telemetry_overflow:
+            return
+        self._telemetry_overflow = True
+        with telemetry.context(obj=self._telemetry_obj) as tel:
+            tel.feature.flow_control_overflow = True
+        telemetry_record = pb.TelemetryRecordRequest(telemetry=self._telemetry_obj)
+        record = self._interface._make_request(telemetry_record=telemetry_record)
+        self._forward_record(record)
+
+    def _pause_marker(self) -> None:
+        self._maybe_update_telemetry()
+        self._send_mark()
+
+    def _write_record(self, record: "pb.Record") -> int:
+        assert self._ds
 
-    def write(self, record):
+        self._record_num += 1
+        proto_util._assign_record_num(record, self._record_num)
+        ret = self._ds.write(record)
+        assert ret is not None
+
+        (_start_offset, end_offset, flush_offset) = ret
+        proto_util._assign_end_offset(record, end_offset)
+        return end_offset
+
+    def _ensure_flushed(self, offset: int) -> None:
+        if self._ds:
+            self._ds.ensure_flushed(offset)
+
+    def _recover_records(self, start: int, end: int) -> None:
+        sender_read = pb.SenderReadRequest(start_offset=start, final_offset=end)
+        # TODO(cancel_paused): implement me
+        # for cancel_id in self._sender_cancel_set:
+        #     sender_read.cancel_list.append(cancel_id)
+        record = self._interface._make_request(sender_read=sender_read)
+        self._ensure_flushed(end)
+        self._forward_record(record)
+
+    def _write(self, record: "pb.Record") -> None:
         if not self._ds:
             self.open()
+        assert self._flow_control
+
+        if not record.control.local:
+            self._write_record(record)
 
+        use_flow_control = (
+            self._settings._network_buffer != 0 and not self._settings._offline
+        )
+        if use_flow_control:
+            self._flow_control.flow(record)
+        elif not self._settings._offline or record.control.always_send:
```

Comment:
could you add a comment about this condition?

Reply:
Added:
```
            # when flow_control is disabled we pass through all records to
            # the sender as long as we are online.   The exception is there
            # are special records that we always pass to the sender
            # (namely the exit record so we can trigger the defer shutdown
            # state machine)
 ```